crates/noirc_evaluator/src/ssa_refactor/ir/instruction.rs

```diff
@@ -277,10 +278,13 @@ impl Instruction {
                 let array = dfg.get_array_constant(*array);
                 let index = dfg.get_numeric_constant(*index);
 
+                // If we are getting a value from a constant array with a constant index
+                // Then we can substitute that value in place of the array_get instruction.
+                // TODO: what is a non-constant array?
```

Comment:
A non-constant array is any value of `Type::Array` for which we don't know the exact values of each element. For example, in `v1 = array_set v0, index Field 1, value Field 7`, `v1` (and v0) are non-constant since we do not know their element values(*). Note that non-constant arrays include more than just "arrays that are parameters to main" because it also includes arrays that may eventually become known constants by later compiler passes.

(*) In the SSA printer if an array is constant the underlying array itself `[0, 1, 2, etc]` is printed, and if it isn't constant just the ID `v1` is printed. So if you see an ID instead of an array value you know it is not constant.